common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/IChallengeHandler.java

```diff
@@ -0,0 +1,27 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.internal.ui.embeddedwebview.challengehandlers;
+
+public interface IChallengeHandler {
```

Comment:
I'm not sure about the value of this interface -- accepts no params and returns nothing.

Would this make sense to provide with a generic type?

As a thought...

```java
public interface IChallengeHandler<Challenge, Response> {

    Response processChallenge(Challenge challenge);

}
```

and then have....

```java
public final class NtlmChallengeHandler implements IChallengeHandler<NtlmChallenge, Void> {

    @Override
    public Void processChallenge(final NtlmChallenge challenge) {
        if (challenge.getHandler().useHttpAuthUsernamePassword() && challenge.getView() != null) {
        
        final String[] haup = challenge.getView().getHttpAuthUsernamePassword(challenge.getHost(), challenge.mRealm);
            
            if (haup != null && haup.length == 2) {
                final String userName = haup[0];
                final String password = haup[1];
                if (userName != null && password != null) {
                    challenge.getHandler().proceed(userName, password);
                }
            }
        } else {
            showHttpAuthDialog();
        }

        return null;
    }

}
```

Reply:
`Response processChallenge(Challenge challenge);`  could not fit into the `ClientCertAuthChallenge` which will take `ClientCertRequest` (pre-defined class in Android SDK) as input. 

And the challenge result is sent to IChallengeCompletionCallback which required the calling Activity.setResult() and Acitivity.finish() in AuthenticationAcitivty of MSAL. That's why I introduced the callback interface here and let the MSAL to have the callback implementation. This is why there is no return object here.

Reply:
Talked to Brian offline. Learned a lot from Brian. I will address this. Thanks @iambmelt